### PR TITLE
Refactor NL server to decouple embeddings model and store

### PR DESCRIPTION
**Pull request by pradh**
The structure now is something like this:

* `Embeddings` - a simple class to encapsulate a pair of `EmbeddingsModel` and `EmbeddingsStore`
  * `EmbeddingsStore`: files in `nl_server/store`
      * Takes in vectors and returns `VarCandidates` list
      * For now just MemoryStore (which was previously "Embeddings" class)
      * Plan is to add LanceDB next...
  * `EmbeddingsModel`: files in `nl_server/model`
      * Takes in multiple queries and returns vectors
      * For now this is just SentenceTransformer
* `EmbeddingsMap` - a global map of index-names to `Embeddings` objects
   * This class knows to construct the stores and models
   * Previously this was confusingly called "EmbeddingsStore"

Here is how the LanceDB integration will look (follow-on PR): https://github.com/datacommonsorg/website/pull/4098

(This PR has ZERO functionality change.)

### CODE REVIEW

File: nl_server/embeddings_map.py [hunk 1]
```diff
@@ -16,21 +16,23 @@
...
 
 
 #
-# A simple wrapper class around multiple embeddings indexes.
+# A map from specific index to embeddings stores and models.
 #
-class Store:
+class EmbeddingsMap:
 
   def __init__(self, indexes: List[EmbeddingsIndex]):
```
Comment (shifucun) [inline nl_server/embeddings_map.py:37]: here assumes relation between index and models. I guess all of those should eventually be config driven.

File: nl_server/embeddings_map.py [hunk 2]
```diff
@@ -47,19 +49,20 @@ def __init__(self, indexes: List[EmbeddingsIndex]):
           default_idx, custom_idx)
```
Comment (shifucun) [inline nl_server/embeddings_map.py:49]: custom dc logic might need to move out.

File: nl_server/embeddings_util.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# limitations under the License.
+"""Helper logic for Embeddings."""
+
+from typing import List
+
+from shared.lib.detected_variables import VarCandidates
+
+# Number of matches to find within the SV index.
+_NUM_SV_INDEX_MATCHES = 40
+# Number of matches to find within the SV index if skipping topics.
```
Comment (shifucun) [inline nl_server/embeddings_util.py:22]: why skipping topics require more matches?

File: nl_server/embeddings_util.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  # Drop the topics from the results.
+  post_trim = []
+  for pre in pre_trim:
+    post = VarCandidates(svs=[], scores=[], sv2sentences={})
+    for i, sv in enumerate(pre.svs):
+      if not sv.startswith(_TOPIC_PREFIX):
+        post.svs.append(sv)
+        post.scores.append(pre.scores[i])
+        post.sv2sentences[sv] = pre.sv2sentences[sv]
+    post_trim.append(post)
```
Comment (shifucun) [inline nl_server/embeddings_util.py:45]: use array filter lambda function?

File: nl_server/model/sentence_transformer.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2023 Google LLC
```
Comment (shifucun) [inline nl_server/model/sentence_transformer.py:1]: 2024

File: nl_server/model/sentence_transformer.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+from typing import List
+
+from sentence_transformers import SentenceTransformer
+import torch
+
+from nl_server import config
+from nl_server import wrapper
+
+
+class SentenceTransformerModel(wrapper.EmbeddingsModel):
```
Comment (shifucun) [inline nl_server/model/sentence_transformer.py:26]: LocalSentenceTransformerModel?

File: nl_server/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+# A simple wrapper around EmbeddingsModel + EmbeddingsStore.
+class Embeddings:
+
+  def __init__(self, model: EmbeddingsModel, store: EmbeddingsStore):
+    self.model: EmbeddingsModel = model
+    self.store: EmbeddingsStore = store
+
+  def search_vars(self,
+                  queries: List[str],
+                  skip_topics: bool = False) -> List[VarCandidates]:
```
Comment (shifucun) [inline nl_server/wrapper.py:65]: In the future, do you think VarCandidates should be reshaped as list of {sv:, score:, sentences[]} as that's a general vector search result?

And the result here should be a map from query to VarCandidate?

**Review by shifucun — COMMENTED** [review]
very nice refactor, the model and embedding layout are becoming more and more clear now!

File: nl_server/store/memory.py [hunk 8]
```diff
@@ -69,74 +54,58 @@ def __init__(self, embeddings_path: str, model: any) -> None:
...
-    top_k = _NUM_SV_INDEX_MATCHES
-    if skip_topics:
-      top_k = _NUM_SV_INDEX_MATCHES_WITHOUT_TOPICS
+  def vector_search(self, query_embeddings: torch.Tensor,
+                    top_k: int) -> List[vars.VarCandidates]:
     hits = semantic_search(query_embeddings,
                            self.dataset_embeddings,
                            top_k=top_k)
 
     # A map from input query -> SV DCID -> matched sentence -> score for that match
```
Comment (chejennifer) [inline nl_server/store/memory.py:66]: update this comment according to the type change?

File: nl_server/store/memory.py [hunk 9]
```diff
@@ -69,74 +54,58 @@ def __init__(self, embeddings_path: str, model: any) -> None:
...
-    query2sv2sentence2score: Dict[str, Dict[str, Dict[str, float]]] = {}
+    query_indexed_sv2sentence2score: List[Dict[str, Dict[str, float]]] = []
     # A map from input query -> SV DCID -> highest matched score
```
Comment (chejennifer) [inline nl_server/store/memory.py:68]: also update this comment?

File: nl_server/store/memory.py [hunk 10]
```diff
@@ -69,74 +54,58 @@ def __init__(self, embeddings_path: str, model: any) -> None:
...
+            query_indexed_sv2sentence2score[i][dcid] = {}
 
           if ent['corpus_id'] >= len(self.sentences):
             continue
           sentence = self.sentences[ent['corpus_id']]
-          query2sv2sentence2score[q][dcid][sentence] = score
+          query_indexed_sv2sentence2score[i][dcid][sentence] = score
 
-    query2result: Dict[str, vars.VarCandidates] = {}
+    results: List[vars.VarCandidates] = []
```
Comment (chejennifer) [inline nl_server/store/memory.py:86]: curious why the change from a list to a dict?

File: nl_server/store/memory.py [hunk 11]
```diff
@@ -69,74 +54,58 @@ def __init__(self, embeddings_path: str, model: any) -> None:
...
 
     # Go over the map and prepare parallel lists of
     # SVs and scores in query2result.
-    for q, sv2score in query2sv2score.items():
+    for sv2score in query_indexed_sv2score:
```
Comment (chejennifer) [inline nl_server/store/memory.py:90]: update the comment

**Review by chejennifer — APPROVED** [review]
nice! Just some small things

File: nl_server/store/memory.py (hunk 8 above)
Comment (pradh) [reply nl_server/store/memory.py]: Done.

File: nl_server/store/memory.py (hunk 9 above)
Comment (pradh) [reply nl_server/store/memory.py]: Done

File: nl_server/store/memory.py (hunk 10 above)
Comment (pradh) [reply nl_server/store/memory.py]: Because the input changed from query string to query embeddings, it seemed more natural to return a list corresponding to the input list.

File: nl_server/embeddings_util.py (hunk 4 above)
Comment (pradh) [reply nl_server/embeddings_util.py]: Done (this is now inlined in wrapper.py)

File: nl_server/wrapper.py (hunk 7 above)
Comment (pradh) [reply nl_server/wrapper.py]: Good idea!

**Review by pradh — COMMENTED** [review]
Thanks for the reviews!

**Review by shifucun — APPROVED** [review]

**Comment by pradh** [comment]
Thanks for the reviews!
